docs/DeveloperGuide.md

```diff
@@ -1,25 +1,283 @@
 # Developer Guide
 
 ## Design & implementation
+<!-- @@author iamchenjiajun -->
+### Architecture
 
-{Describe the design and implementation of the product. Use UML diagrams and short code snippets where applicable.}
+##### Architecture Diagram
+Below is an architecture diagram of termiNus.
+
+![ArchitectureDiagram](./images/ArchitectureDiagram.png)
+
+- `Duke` is the main object of the program and handles all the logic and models related to the program.
+- `Ui` is the main object that provides an interface between `Duke` and the user.
+- `Command` represents a command that is provided by the user. `Ui` reads the command and it is sent to `Parser` to create a new `Command` object.
+- `Duke` executes the command and shows the output to the user using `Ui`.
+- `Command` object modifies the state of `ItemList`, which consists of multiple lists for different types of `Items`.
+- `Storage` takes the state of `ItemList` and stores it to file.
+
+##### Sequence Diagram
+Below is a sequence diagram of how the main program functions.
+
+![DukeSequenceDiagram](./images/DukeSequenceDiagram.png)
+
+- First, the `main` function of the `Duke` class creates an instance of `Duke`.
+- During the instantiation of `Duke`, a `Storage` object is created.
+- `Duke` creates and loads the state of `ItemList` from file by calling a function from `Storage`.
+- After `Duke` is initialized, the `Duke` class calls the `run()` method of `Duke`.
+- `Duke` calls methods from `Ui` class and shows messages to the user.
+- `Duke` reads commands from the user using the `Ui` class (which acts as an interface between `Duke` and the user).
+- `Command` object is returned to `Duke` which is executed.
+- `Command` object interacts with `ItemList` and changes its state.
+- `Storage` saves the state of `ItemList` to file.
+- `Duke` continues reading commands until a `ByeCommand` is generated by the user.
+
+<!-- @@author -->
+### Ui component
+
+<!-- @@author iamchenjiajun -->
+### Parser component
+
+The `Parser` class is a class forming part of the logic of termiNus. The `Parser` parses user commands and
+returns a `Command` subclass which corresponds to the full command with arguments.
+
+#### High level description
+
+The `Parser.parse()` method takes in `fullCommand` as an argument, which is the full command entered by the user.
+
+Example commands:
+- `add task tP meeting c/cs2113 p/0`
+- `add module CS2113 mc/4 ay/2021S1 g/A`
+
+The `fullCommand` is composed of several parts, which consists of a root command (`add`), description (`module CS2113`)
+and optional arguments (`mc/4 ay/2021S1 g/A`).
+
+An optional argument consists of 2 parts, which is delimited by a forward slash. In the example above, there are 3 optional
+arguments, which are `mc/4`, `ay/2021S1` and `g/A`. Each optional argument can be represented in this form: `<key>/<value>`.
```

Comment:
This is a good explanation and makes it clear, especially `<key>/<value>`. Maybe it would be worth it to explicitly list out the keys as being `mc`, `ay` and `g` and what they each mean even if it can be inferred.